File: Python/156-binaryTreeUpsideDown.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None
    
# Chris:TODO::write following again, so sleepy.
class Solution:
    # @param root, a tree node
    # @return root of the upside down tree
    def upsideDownBinaryTree(self, root):
        parent = None
        parent_left = None
        while root:
              left = root.left
              right = root.right
              root.left = parent
              parent = root
              root.right = parent_left
              parent_left = right
              root = left
        return parent

if __name__ == '__main__':
   s = Solution()
   
   test = TreeNode(1)
   test.left = TreeNode(2)
   test.right = TreeNode(3)
   test.left.left = TreeNode(4)
   test.left.right = TreeNode(5)

   print s.upsideDownBinaryTree(test)

File: C++/156-BinaryTreeUpsideDown.cpp
#include "include/common.h"

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

class Solution {
public:
    TreeNode* upsideDownBinaryTree(TreeNode* root) {
              struct TreeNode *parent = nullptr;
              struct TreeNode *parentLeft = nullptr;
              struct TreeNode *rootLeft, *rootRight;
              while (root) {
                     rootLeft  = root->left;
                     rootRight = root->right;
                     root->left = parentLeft;
                     root->right = parent;
                     parent = root;
                     parentLeft = rootRight;
                     root = rootLeft;
              };
              return parent;
    }
};